dask/array/_shuffle.py

```diff
@@ -0,0 +1,199 @@
+    """
+    Reorders one dimensions of a Dask Array based on an indexer.
+
+    The indexer defines a list of positional groups that will end up in the same chunk
+    together. A single group is in at most one chunk on this dimension, but a chunk
+    might contain multiple groups to avoid fragmentation of the array.
+
+    The algorithm tries to balance the chunksizes as much as possible to ideally keep the
+    number of chunks consistent or at least manageable.
+
+    Parameters
+    ----------
+    x: dask array
+        Array to be shuffled.
+    indexer:  list[list[int]]
+        The indexer that determins which elements along the dimension will end up in the
+        same chunk. Multiple groups can be in the same chunk to avoid fragmentation, but
+        each group will end up in exactly one chunk.
+    axis: int
+        The axis to shuffle along.
+
+    Examples
+    --------
+    >>> import dask.array as da
+    >>> import numpy as np
+    >>> arr = np.array([[1, 2, 3, 4, 5, 6, 7, 8], [9, 10, 11, 12, 13, 14, 15, 16]])
+    >>> x = da.from_array(arr, chunks=(2, 4))
+
+    Separate the elements in different groups.
+
+    >>> y = x.shuffle([[6, 5, 2], [4, 1], [3, 0, 7]], axis=1)
+
+    The shuffle algorihthm will combine the first 2 groups into a single chunk to keep
+    the number of chunks small.
+
+    The tolerance of increasing the chunk size is controlled by the configuration
+    "array.shuffle.chunksize-tolerance". The default value is 1.25.
+
+    >>> y.chunks
+    ((2,), (5, 3))
+
+    The array was reordered along axis 1 according to the positional indexer that was given.
+
+    >>> y.compute()
+    array([[ 7,  6,  3,  5,  2,  4,  1,  8],
+           [15, 14, 11, 13, 10, 12,  9, 16]])
+    """
+    if np.isnan(x.shape).any():
+        raise ValueError(
+            f"Shuffling only allowed with known chunk sizes. {unknown_chunk_message}"
+        )
+    assert isinstance(axis, int), "axis must be an integer"
+
+    token = tokenize(x, indexer, axis)
+    out_name = f"shuffle-{token}"
+
+    chunks, layer = _shuffle(x.chunks, indexer, axis, x.name, out_name, token)
+    graph = HighLevelGraph.from_collections(out_name, layer, dependencies=[x])
+
+    return Array(graph, out_name, chunks, meta=x)
+
+
+def _shuffle(chunks, indexer, axis, in_name, out_name, token):
+    if not isinstance(indexer, list) or not all(isinstance(i, list) for i in indexer):
+        raise ValueError("indexer must be a list of lists of positional indices")
+
+    if not axis <= len(chunks):
+        raise ValueError(
+            f"Axis {axis} is out of bounds for array with {len(chunks)} axes"
+        )
+
+    if max(map(max, indexer)) >= sum(chunks[axis]):
+        raise IndexError(
+            f"Indexer contains out of bounds index. Dimension only has {sum(chunks[axis])} elements."
+        )
+
+    chunksize_tolerance = config.get("array.shuffle.chunksize-tolerance")
+    average_chunk_size = int(
```

Comment:
nit: Rename since this is the average multiplied by the tolerance.

```suggestion
    chunk_size_limit = int(
```